osquery/tables/networking/aws_metadata.h

```diff
@@ -0,0 +1,113 @@
+/*
+ *  Copyright (c) 2016-present, Jean-Francois Dive <jfdive@gmail.com>
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#pragma once
+
+#include <string>
+
+#include "include/osquery/tables.h"
+
+namespace osquery {
+namespace tables {
+
+/**
+ * @brief SuperClass for AWS metadata accessors
+ */
+class AwsData {
+  protected:
+  /**
+   * @brief SQL type for the value
+   */
+  ColumnType sqlType;
+
+  /**
+   * @brief SQL column name
+   */
+  std::string fieldName;
+
+  /**
+   * @brief API URL
+   */
+  std::string subUrl;
+
+  /**
+   * @brief HTTP get the data
+   *
+   * @return osquery Status
+   */
+  std::string DoGet();
+
+  public:
+  /**
+   * @brief ctor
+   */
+  AwsData(ColumnType in_sqlType, std::string in_fieldName, std::string in_subUrl) {
```

Comment:
If the arguments include non-references you're invoking the copy constructor for that type (in this case `std::string`). This is fine if you want to include a copy locally, but there's added work needed.

For non POD types, like a `std::string` this constructor will invoke the member variable initialization twice (so to speak), once when the object is constructed, then again for each when their copy assignment constructor is called in the constructor body.

If you write something like:
```cpp
AwsData(ColumnType _sqlType, std::string _fieldName, std::string _subUrl)
  : sqlType(_sqlType), fieldName(std::move(_fieldName), subUrl(std::move(_subUrl)) {}
```